x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredModelSnapshotsRemover.java

```diff
@@ -175,19 +191,28 @@ protected void removeDataBefore(
             public void onResponse(SearchResponse searchResponse) {
                 long nextToKeepMs = deleteAllBeforeMs;
                 try {
-                    List<ModelSnapshot> modelSnapshots = new ArrayList<>();
+                    List<JobSnapshotId> snapshotIds = new ArrayList<>();
                     for (SearchHit hit : searchResponse.getHits()) {
-                        ModelSnapshot modelSnapshot = ModelSnapshot.fromJson(hit.getSourceRef());
-                        long timestampMs = modelSnapshot.getTimestamp().getTime();
+                        String timestamp = stringFieldValueOrNull(hit, ModelSnapshot.TIMESTAMP.getPreferredName());
+                        if (timestamp == null) {
+                            continue;
```

Comment:
Maybe we should log an error here, as I don't think this should never happen unless a document is corrupt.  So if it _does_ happen it would be useful to know about when debugging why results haven't been deleted.